waspc/e2e-test/Main.hs

```diff
@@ -0,0 +1,54 @@
+import Control.Monad (unless)
+import System.Directory (getCurrentDirectory)
+import System.FilePath (joinPath, takeFileName, (</>))
+import System.Info
+import Test.Tasty (TestTree, defaultMain, testGroup)
+import Util
+  ( runGoldenTest,
+    waspCliNew,
+    waspCliNewAndCompile,
+    waspCliNewAndCompileAndMigrate,
+  )
+
+-- TODO: Put `cwd` into ReaderT Monad?
+
+-- TODO: Switch over from FilePath to StrongPath eventually and use helper like so:
+--
+-- data ProjectRoot
+-- getProjectRootPath :: IO (Path Abs (Dir ProjectRoot))
+-- getProjectRootPath = getCurrentDirectory
+
+type GoldenTest = FilePath -> IO TestTree
+
+main :: IO ()
+main = do
+  if os == "mingw32"
+    then putStrLn "Skipping end-to-end tests on Windows due to tests using *nix-only commands"
+    else do
+      -- Stack/Cabal launch `stack test` from root of project, so this should always be some absolute path to waspc
+      cwd <- getCurrentDirectory
+      unless (takeFileName cwd == "waspc") (error "Expecting test process to be invoked from waspc dir")
+      defaultMain =<< tests cwd
+
+tests :: GoldenTest
+tests cwd =
+  testGroup "All Golden Dir Tests"
+    <$> sequence
+      [ waspNew cwd,
+        waspCompile cwd,
+        waspMigrate cwd
+      ]
+
+waspNew :: GoldenTest
+waspNew cwd =
+  runGoldenTest cwd "waspNew" waspCliNew
```

Comment:
Extracting functions into another module hides what the test actually does from the place where the test gets run. I'm a bit concerned about that.

Given the amount of repetition left, I'd be inclined to investigate whether we can declare tests as a list:

```haskell
[("waspNew", waspCliNew), ("waspCompile", waspCliNewAndCompile)]
```

and have a test runner.

Reply:
I think we were able to chat this one through on the call